### ISSUE TITLE
The CharSequence hack in FunctionEmitter is unsound

### ISSUE THREAD

**Issue by sjrd** [scala-js/scala-js#4408]
It turns out [the `CharSequence` hack in `FunctionEmitter`](https://github.com/scala-js/scala-js/blob/b3b964ccdf2dd79038eaca98ce89fa9f8fbdcf8a/linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala#L2184-L2204) is unsound. Who would have thought?

The following program reproduces the issue:
```scala
package helloworld

import scala.scalajs.js

object HelloWorld {
  def main(args: Array[String]): Unit = {
    test("hello")
    test(new java.lang.StringBuilder("hello"))
  }

  @noinline def test(sequence: CharSequence): Unit = {
    js.Dynamic.global.console.log(sequence.subSequence(2, 4).toString())
  }
}
```
The issue cannot be reproduced in the test suite because there are too many implementations of `CharSequence`.

The output is
```
ll
C:\Users\sjrdo\Documents\Projets\scalajs\examples\helloworld\.2.12\target\scala-2.12\helloworld-fastopt\main.js:1518
  return $as_T($thiz.substring(beginIndex, endIndex))
                     ^

TypeError: $thiz.substring is not a function
    at $f_T__substring__I__I__T (C:\Users\sjrdo\Documents\Projets\scalajs\examples\helloworld\.2.12\target\scala-2.12\helloworld-fastopt\file:\C:\Users\sjrdo\Documents\Projets\scalajs\javalanglib\src\main\scala\java\lang\_String.scala:373:17)
    at $c_Lhelloworld_HelloWorld$.test__jl_CharSequence__V (C:\Users\sjrdo\Documents\Projets\scalajs\examples\helloworld\.2.12\target\scala-2.12\helloworld-fastopt\file:\C:\Users\sjrdo\Documents\Projets\scalajs\javalanglib\src\main\scala\java\lang\_String.scala:364:14)
    at $c_Lhelloworld_HelloWorld$.main__AT__V (C:\Users\sjrdo\Documents\Projets\scalajs\examples\helloworld\.2.12\target\scala-2.12\helloworld-fastopt\file:\C:\Users\sjrdo\Documents\Projets\scalajs\examples\helloworld\src\main\scala\helloworld\HelloWorld.scala:24:9)
    at $s_Lhelloworld_HelloWorld__main__AT__V (C:\Users\sjrdo\Documents\Projets\scalajs\examples\helloworld\.2.12\target\scala-2.12\helloworld-fastopt\file:\C:\Users\sjrdo\Documents\Projets\scalajs\examples\helloworld\src\main\scala\helloworld\HelloWorld.scala:20:8)
    at C:\Users\sjrdo\Documents\Projets\scalajs\examples\helloworld\.2.12\target\scala-2.12\helloworld-fastopt\main.js:1846:1
    at C:\Users\sjrdo\Documents\Projets\scalajs\examples\helloworld\.2.12\target\scala-2.12\helloworld-fastopt\main.js:1847:4
    at Script.runInThisContext (vm.js:132:18)
    at Object.runInThisContext (vm.js:309:38)
    at [stdin]:8:25
    at Script.runInThisContext (vm.js:132:18)
```
The first call succeeds, but the second one fails. This happens because we attempt to call `$f_T__substring__I__I__T` on the `j.l.StringBuilder`. The compiled code of `test` is:
```js
$c_Lhelloworld_HelloWorld$.prototype.test__jl_CharSequence__V = (function(sequence) {
  var $$x1 = console;
  var this$1 = $f_T__substring__I__I__T(sequence, 2, 4);
  $$x1.log(this$1)
});
```
How does that happen? Because of multi-inlining. When the optimizer sees the call
```scala
sequence.subSequence;I;I;Ljava.lang.CharSequence(2, 4)
```
it looks for possible implementations of `subSequence`. There are exactly 2: one in `String`, and one in `StringBuilder`. It turns out that both those implementations delegate to a `substring` method, with the same arguments, returning a `String`. That triggers multi-inlining.

Since we're multi-inlining, the receiver type is (correctly) not refined, and we end up with
```scala
sequence.substring;I;I;Ljava.lang.String(2, 4)
```
where `sequence` is a `CharSequence`.

When we get to the `FunctionEmitter`, this enters the `CharSequence` hack, which assumes that this must be a call on a `String`, and emits JS code to call the hijacked class method `$f_T__substring__I__I__T`.

---

In theory, this can happen with any other hijacked class, but in practice it doesn't. That's because, for all the methods in all the other hijacked classes, there is a second hijacked class where that method also exists and does something different. Since all hijacked classes are instantiated, multi-inlining never happens.

---

How to fix it? Well, first we need to remove the hack, which means that `FunctionEmitter` would emit a call to a dispatcher if it ends up in the above situation. That's still no good because we don't have dispatchers for the dozens of methods in `String`. There are two ways out of this:
* the workaround way: prevent multi-inlining when one of the targets is in a hijacked class
* the real fix way: actually declare dispatchers for all the methods of all hijacked classes, including `String`.

The second way is better, but we cannot rely on having a hard-coded list of all the methods of `String`, like we currently do for the dispatcher. So we would actually have to dynamically list all reachable methods in hijacked classes, and generate dispatchers for those.

**Comment by gzm0** [comment]
Does this mean less `EmitterNames`? :sparkles:

**Comment by gzm0** [comment]
Hmmm... Will we really gain anything by allowing multi-inlining in these scenarios if it means we need dispatchers for *all* methods of HijackedClasses (not just overridden ones)? (by "gain" I mean some speed or size improvement).

**Comment by sjrd** [comment]
> Does this mean less `EmitterNames`? ✨

Yes. A lot less.

> Hmmm... Will we really gain anything by allowing multi-inlining in these scenarios if it means we need dispatchers for _all_ methods of HijackedClasses (not just overridden ones)? (by "gain" I mean some speed or size improvement).

A little bit, yes, though nothing fundamental. In the scenario when we prevent multi-inlining, we would get something like:
* call `$dp_subSequence`
* that dispatches between `String` and non-`String`
* If `String`, call `$f__T__subSequence`, which calls `$f__T__substring`
* If non-`String`, call `x.subSequence`, which dynamically resolves, and there we call `this.substring`

In the scenario where we allow multi-inlining and we have dispatchers for all methods, it goes like this:
* call `$dp_substring` (because it was multi-inline)
* that dispatches between `String` and non-`String`
* If `String`, call `$f__T__substring`
* If non-`String` call `x.substring`

In both the `String` and non-`String` case we have avoided one method call (good for speed) and the method we have avoided can be dce'ed (good for code size). We've also traded one dispatcher for another in the first step, but that has 0 impact on speed or code size.
